File: dotnet/src/Microsoft.Agents.Protocol.Validation/ValidationError.cs
namespace Microsoft.Agents.Protocol.Validation;

/// <summary>
/// Represents a validation error with structured information.
/// </summary>
public class ValidationError
{
    /// <summary>
    /// Gets or sets the error code (e.g., "MSG-001", "REL-002").
    /// </summary>
    public string Code { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the human-readable error message.
    /// </summary>
    public string Message { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the field or property name that failed validation.
    /// </summary>
    public string? Field { get; set; }

    /// <summary>
    /// Gets or sets the severity of the validation error.
    /// </summary>
    public ValidationSeverity Severity { get; set; } = ValidationSeverity.Error;

    /// <summary>
    /// Gets or sets additional context information for the error.
    /// </summary>
    public Dictionary<string, object>? Context { get; set; }

    /// <summary>
    /// Creates a new validation error.
    /// </summary>
    public ValidationError()
    {
    }

    /// <summary>
    /// Creates a new validation error with the specified code and message.
    /// </summary>
    public ValidationError(string code, string message, string? field = null, ValidationSeverity severity = ValidationSeverity.Error)
    {
        Code = code;
        Message = message;
        Field = field;
        Severity = severity;
    }

    /// <summary>
    /// Returns a string representation of this validation error.
    /// </summary>
    public override string ToString()
    {
        var fieldPart = !string.IsNullOrEmpty(Field) ? $" [{Field}]" : "";
        return $"{Code}{fieldPart}: {Message}";
    }
}

/// <summary>
/// Severity levels for validation errors.
/// </summary>
public enum ValidationSeverity
{
    /// <summary>
    /// Informational message.
    /// </summary>
    Info,

    /// <summary>
    /// Warning that should be addressed.
    /// </summary>
    Warning,

    /// <summary>
    /// Error that must be fixed.
    /// </summary>
    Error
}

File: typescript/packages/agents-protocol-validation/src/core/ValidationError.ts
/**
 * Severity levels for validation errors.
 */
export enum ValidationSeverity {
  Info = 'info',
  Warning = 'warning',
  Error = 'error'
}

/**
 * Represents a validation error with structured information.
 */
export class ValidationError {
  /**
   * The error code (e.g., "MSG-001", "REL-002").
   */
  public readonly code: string;

  /**
   * The human-readable error message.
   */
  public readonly message: string;

  /**
   * The field or property name that failed validation.
   */
  public readonly field?: string;

  /**
   * The severity of the validation error.
   */
  public readonly severity: ValidationSeverity;

  /**
   * Additional context information for the error.
   */
  public readonly context?: Record<string, unknown>;

  constructor(
    code: string,
    message: string,
    field?: string,
    severity: ValidationSeverity = ValidationSeverity.Error,
    context?: Record<string, unknown>
  ) {
    this.code = code;
    this.message = message;
    this.field = field;
    this.severity = severity;
    this.context = context;
  }

  /**
   * Returns a string representation of this validation error.
   */
  toString(): string {
    const fieldPart = this.field ? ` [${this.field}]` : '';
    return `${this.code}${fieldPart}: ${this.message}`;
  }
}
